internal/contentmanager/cache/cache.go

```diff
@@ -0,0 +1,240 @@
+	}
+
+	if err := c.removeStaleSources(gvkSet); err != nil {
+		return fmt.Errorf("removing stale sources: %w", err)
+	}
+	return c.startNewSources(ctx, gvkSet, watcher)
+}
+
+func (c *cache) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	errs := []error{}
+	for _, source := range c.sources {
+		if err := source.Close(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+func (c *cache) startNewSources(ctx context.Context, gvks sets.Set[schema.GroupVersionKind], watcher Watcher) error {
+	cacheGvks := c.getCacheGVKs()
+	gvksToCreate := gvks.Difference(cacheGvks)
+
+	type startResult struct {
+		source CloserSyncingSource
+		gvk    schema.GroupVersionKind
+		err    error
+	}
+	startResults := make(chan startResult)
+	wg := sync.WaitGroup{}
+	for _, gvk := range gvksToCreate.UnsortedList() {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			source, err := c.startNewSource(ctx, gvk, watcher)
+			startResults <- startResult{
+				source: source,
+				gvk:    gvk,
+				err:    err,
+			}
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(startResults)
+	}()
+
+	sourcesErrors := []error{}
+	for result := range startResults {
+		if result.err != nil {
+			sourcesErrors = append(sourcesErrors, result.err)
+			continue
+		}
+
+		err := c.addSource(result.gvk, result.source)
+		if err != nil {
+			// If we made it here then there is a logic error in
+			// calculating the diffs between what is currently being
+			// watched by the cache
+			panic(err)
+		}
+	}
+
+	return errors.Join(sourcesErrors...)
+}
+
+func (c *cache) startNewSource(ctx context.Context, gvk schema.GroupVersionKind, watcher Watcher) (CloserSyncingSource, error) {
+	s, err := c.sourcerer.Source(gvk, c.owner, func(ctx context.Context) {
+		err := c.removeSource(gvk)
+		if err != nil {
+			logr := log.FromContext(ctx)
+			logr.Error(err, "managed content cache postSyncError removing source failed", "gvk", gvk)
+		}
+	})
+	if err != nil {
+		return nil, fmt.Errorf("getting source: %w", err)
+	}
+
+	err = watcher.Watch(s)
+	if err != nil {
+		return nil, fmt.Errorf("establishing watch for GVK %q: %w", gvk, err)
+	}
+
+	syncCtx, syncCancel := context.WithTimeout(ctx, c.syncTimeout)
+	defer syncCancel()
+	err = s.WaitForSync(syncCtx)
+	if err != nil {
+		return nil, fmt.Errorf("waiting for sync: %w", err)
+	}
+
+	return s, nil
+}
+
+func (c *cache) addSource(gvk schema.GroupVersionKind, source CloserSyncingSource) error {
+	if _, ok := c.sources[gvk]; !ok {
+		c.sources[gvk] = source
+		return nil
+	}
+	return errors.New("source already exists")
+}
+
+func (c *cache) removeStaleSources(gvks sets.Set[schema.GroupVersionKind]) error {
+	cacheGvks := c.getCacheGVKs()
+	removeErrs := []error{}
+	gvksToRemove := cacheGvks.Difference(gvks)
+	for _, gvk := range gvksToRemove.UnsortedList() {
+		err := c.removeSource(gvk)
+		if err != nil {
+			removeErrs = append(removeErrs, err)
+		}
+	}
+	return errors.Join(removeErrs...)
```

Comment:
Similar reasoning as other comments: sort these `removeErrs`?

Reply:
Should be sorted by error string in https://github.com/operator-framework/operator-controller/pull/1119/commits/f6f277035a8331898fd304504b30660905797406